### DISCUSSION TITLE
Reading config programatically

### DISCUSSION THREAD

**Discussion by vemv** [clj-kondo/clj-kondo#1555] [Q&A]
## Context

https://github.com/clojure-emacs/refactor-nrepl/issues/359#issuecomment-1019639865

## Q

What's the recommended way to programatically read clj-kondo confifg from JVM clojure? I'd assume it's a composition of maps rather than a single `slurp`ing.

(I have various code that currently slurps I think - good chance to revisit my assumptions)

**Comment by borkdude** [comment]
There is the `resolve-config` function for this: https://github.com/clj-kondo/clj-kondo/blob/504077db186c00e09ec16bd85c01ebafa2cb86f1/src/clj_kondo/core.clj#L219

**Reply by vemv** [reply]
Thanks!

A related question is whether this helper could be offered in a way that doesn't imply pulling other dependencies.

CIDER projects are ideally as lean as possible, since they must go through [mranderson](https://github.com/benedekfazekas/mranderson). The less stuff we feed to it the better.

(clojure-lsp generously gave this treatment to its own codebase recently)

**Reply by borkdude** [reply]
We could pull this logic out into a library. I would do this if there was a serious chance that this library would be indeed used elsewhere outside of clj-kondo, as it takes time to do this.

**Reply by vemv** [reply]
Cheers. I can guarantee that it will make it into refactor-nrepl at the very least, in a timely manner.

It doesn't seem contrived to read this stuff from Eastwood either.

**Reply by borkdude** [reply]
Can you explain more about how you would use the library? The output is a bunch of EDN, usually just the `.clj-kondo/config.edn` (so you could start with just prototyping with that) and then, how would you use it in clj-refactor etc?

**Reply by vemv** [reply]
Good thinking, I'll attach a draft PR against refactor-nrepl with said prototype during the week.

**Reply by vemv** [reply]
Done! https://github.com/clojure-emacs/refactor-nrepl/pull/361

The effect is that refactor-nrepl would abstain from cleaning up unused namespaces when kondo config indicates so.

**Reply by borkdude** [reply]
Hi @vemv - I'm not sure what the relation of `:unresolved-namespace` and cleaning unused namespaces is. Could you clarify this using an example? In clj-kondo `:unresolved-namespace :exclude` is used to suppress warnings about missing requires, e.g.:

```
(clojure.string/join ":" [1 2 3])
```

without:

```
(require '[clojure.string])
```

**Reply by vemv** [reply]
whoops, surely I picked the wrong piece of config

**Reply by vemv** [reply]
(Updated the PR a couple fixes)
